File: sdks/python/client/src/max_client/runtime/base_api_client.py
from __future__ import annotations

import json
from typing import Any, TypedDict
from urllib.parse import urlencode, urljoin
from urllib.request import Request, urlopen

from .api_error import MaxApiError, parse_error_payload


class ClientConfig(TypedDict, total=False):
    base_url: str
    access_token: str
    authorization_scheme: str
    default_headers: dict[str, str]


class RequestOptions(TypedDict, total=False):
    headers: dict[str, str]


class BaseApiClient:
    def __init__(self, config: ClientConfig | None = None) -> None:
        config = config or {}
        self.base_url = config.get("base_url", "https://api.max.ru")
        self.access_token = config.get("access_token")
        self.authorization_scheme = config.get("authorization_scheme", "Bearer")
        self.default_headers = config.get("default_headers", {})

    def request(
        self,
        *,
        method: str,
        path: str,
        query: dict[str, Any] | None = None,
        body: Any | None = None,
        options: RequestOptions | None = None,
    ) -> Any:
        url = urljoin(self.base_url, path)
        query_string = build_query_string(query)
        if query_string:
            url = f"{url}?{query_string}"

        headers = dict(self.default_headers)
        if self.access_token:
            headers["Authorization"] = f"{self.authorization_scheme} {self.access_token}"

        if options and options.get("headers"):
            headers.update(options["headers"])

        data: bytes | None = None
        if body is not None:
            headers.setdefault("Content-Type", "application/json")
            data = json.dumps(body).encode("utf-8")

        request = Request(url=url, method=method, data=data, headers=headers)

        try:
            with urlopen(request) as response:
                payload = response.read().decode("utf-8")
                if not payload:
                    return None
                return json.loads(payload)
        except Exception as exc:
            status = getattr(exc, "code", 0)
            body_text = ""
            if hasattr(exc, "read"):
                body_text = exc.read().decode("utf-8")
            payload = parse_error_payload(body_text)
            raise MaxApiError(method=method, path=path, status=status, body_text=body_text, payload=payload) from exc


def build_query_string(query: dict[str, Any] | None) -> str:
    if not query:
        return ""

    items: list[tuple[str, str]] = []
    for key, value in query.items():
        if value is None:
            continue

        if isinstance(value, list):
            for item in value:
                items.append((key, serialize_query_value(item)))
            continue

        items.append((key, serialize_query_value(value)))

    return urlencode(items)


def serialize_query_value(value: Any) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)

File: sdks/typescript/client/runtime/base-api-client.ts
import { MaxApiError, parseMaxErrorPayload } from "./api-error.js";

export interface ClientConfig {
  baseUrl?: string;
  accessToken?: string;
  authorizationScheme?: string;
  fetch?: typeof fetch;
  defaultHeaders?: Record<string, string>;
}

export interface RequestOptions {
  signal?: AbortSignal;
  headers?: Record<string, string>;
}

export interface ApiRequest {
  method: string;
  path: string;
  query?: Record<string, unknown>;
  body?: unknown;
  options?: RequestOptions;
}

export class BaseApiClient {
  protected readonly baseUrl: string;

  private readonly accessToken?: string;
  private readonly authorizationScheme: string;
  private readonly fetchImpl: typeof fetch;
  private readonly defaultHeaders: Record<string, string>;

  constructor(config: ClientConfig = {}) {
    this.baseUrl = config.baseUrl ?? "https://api.max.ru";
    this.accessToken = config.accessToken;
    this.authorizationScheme = config.authorizationScheme ?? "Bearer";
    this.fetchImpl = config.fetch ?? fetch;
    this.defaultHeaders = config.defaultHeaders ?? {};
  }

  protected async request<T>(request: ApiRequest): Promise<T> {
    const url = new URL(request.path, this.baseUrl);
    appendQueryParams(url, request.query);

    const headers = new Headers(this.defaultHeaders);
    if (this.accessToken) {
      headers.set("Authorization", `${this.authorizationScheme} ${this.accessToken}`);
    }

    for (const [key, value] of Object.entries(request.options?.headers ?? {})) {
      headers.set(key, value);
    }

    let body: string | undefined;
    if (request.body !== undefined) {
      if (!headers.has("Content-Type")) {
        headers.set("Content-Type", "application/json");
      }

      body = JSON.stringify(request.body);
    }

    const response = await this.fetchImpl(url, {
      method: request.method,
      headers,
      signal: request.options?.signal,
      body,
    });

    if (!response.ok) {
      const bodyText = await response.text();
      const payload = parseMaxErrorPayload(bodyText);
      throw new MaxApiError({
        method: request.method,
        path: request.path,
        status: response.status,
        bodyText,
        payload,
      });
    }

    if (response.status === 204) {
      return undefined as T;
    }

    const text = await response.text();
    if (!text) {
      return undefined as T;
    }

    return JSON.parse(text) as T;
  }
}

function appendQueryParams(url: URL, query: Record<string, unknown> | undefined): void {
  if (!query) {
    return;
  }

  for (const [key, value] of Object.entries(query)) {
    if (value === undefined || value === null) {
      continue;
    }

    if (Array.isArray(value)) {
      for (const item of value) {
        url.searchParams.append(key, serializeQueryValue(item));
      }
      continue;
    }

    url.searchParams.set(key, serializeQueryValue(value));
  }
}

function serializeQueryValue(value: unknown): string {
  if (typeof value === "boolean") {
    return value ? "true" : "false";
  }

  return String(value);
}
